### PR TITLE
Fixed database declaration error and domain specific settings form alter

### LINKED ISSUE #25: $database declaration in settings.php must use expanded format like D7
After installing the module the error
```
Domain access failed to load during phase: bootstrap include. Please check your settings.php file and site configuration
```
appears on every page load.

@rudy880719 and @sudipto68 have identified that the problem is the syntax used for defining the $database. See [rudy880719/domain#4](https://github.com/rudy880719/domain/issues/4) It fails if the sort syntax is used:
`$database = 'mysql://user:pass@localhost/database_name';`
 
Using the extended syntax like D7 works:
 ```
$databases['default']['default'] = array(
  'driver' => 'mysql',
  'database' => 'database_name',
  'username' => 'user',
  'password' => 'pass',
  'host' => 'localhost',
  'port' => '',
  'charset' => 'utf8mb4',
  'collation' => 'utf8mb4_unicode_ci',
);
```
It is odd that one method of declaring the database works and the other does not.

### PR DESCRIPTION
**Pull request by rudy880719**
Fixes #25

### CODE REVIEW

**Comment by rudy880719** [comment]
Fixes https://github.com/backdrop-contrib/domain/issues/25

File: domain_settings/config/domain_settings.settings.json
Comment (yorkshire-pudding) [inline domain_settings/config/domain_settings.settings.json:1]: Is this change relevant for this issue? I can't see how it relates.

```
-    "domain_settings_behavior": 0,
-    "domain_settings_form_visibility": "DOMAIN_SETTINGS_SHOW_EXCEPT_LISTED",
+    "domain_settings_behavior": 1,
+    "domain_settings_form_visibility": 0,
```

File: settings.inc [hunk 1]
```diff
@@ -23,6 +23,31 @@ include 'domain.bootstrap.inc';
...
+      'prefix' => !empty($database_prefix) ? $database_prefix : '',
+    );
+    $required_parts = array('driver', 'database', 'username', 'host');
+    foreach ($required_parts as $key) {
+      if (!$databases['default']['default'][$key]) {
+        $missing_field = ($key == 'database') ? 'name' : $key;
+        trigger_error('Could not find a value for the database "' . $missing_field . '" setting. Please check the $database setting in settings.php.', E_USER_ERROR);
+      }
+    }
+  }
```
Comment (yorkshire-pudding) [inline settings.inc]:
```suggestion
  $database_is_set = FALSE;
  // Database connection settings have been provided in the form of a connection
  // string.
  if (!empty($database) && is_string($database)) {
    $default_db_string = 'mysql://user:pass@localhost/database_name';
    if ($database !== $default_db_string) {
      // Convert simplified database settings string into the full databases
      // array.
      $database_parts = parse_url($database);
      if (!$database_parts) {
        trigger_error('The database setting could not be parsed. Please check the $database setting in settings.php.', E_USER_ERROR);
      }
      $databases['default']['default'] = array(
        'driver' => $database_parts['scheme'],
        // Remove leading slash.
        'database' => rawurldecode(substr($database_parts['path'], 1)),
        'username' => isset($database_parts['user']) ? rawurldecode($database_parts['user']) : '',
        'password' => isset($database_parts['pass']) ? rawurldecode($database_parts['pass']) : '',
        'host' => $database_parts['host'],
        'port' => isset($database_parts['port']) ? $database_parts['port'] : NULL,
        'prefix' => !empty($database_prefix) ? $database_prefix : '',
      );
      $database_is_set = TRUE;
    }
  }
  // Database connection settings have been provided in the form of a simplified
  // array.
  if (!empty($database) && is_array($database)) {
    $default_db_array = array(
      'database' => 'database_name',
      'username' => 'user',
      'password' => 'pass',
      'host' => 'localhost',
    );
    if ($database != $default_db_array) {
      // Specify default values.
      $database += array(
        'prefix' => '',
        'driver' => 'mysql',
        'port' => NULL,
        'charset' => 'utf8mb4',
        'collation' => 'utf8mb4_general_ci',
      );
      $databases['default']['default'] = $database;
      $database_is_set = TRUE;
    }
  }
  if ($database_is_set) {
    // Check that all required parts are included.
    $required_parts = array('driver', 'database', 'username', 'host');
    foreach ($required_parts as $key) {
      if (!$databases['default']['default'][$key]) {
        $missing_field = ($key == 'database') ? 'name' : $key;
        trigger_error('Could not find a value for the database "' . $missing_field . '" setting. Please check the $database setting in settings.php.', E_USER_ERROR);
      }
    }
  }
```

File: settings.inc [hunk 2]
```diff
@@ -23,6 +23,31 @@ include 'domain.bootstrap.inc';
 // can explode rather maginificently. Using a try/catch, we can at least fail
 // silently.
 try {
+  global $databases, $settings;
```
Comment (yorkshire-pudding) [inline settings.inc]: I don't think `$settings` is needed.  I tested without it and didn't notice any difference. But am I missing something?

```suggestion
  global $databases;
```

**Review by yorkshire-pudding — CHANGES_REQUESTED** [review]
@rudy880719 
Thank you for this.  I had tried this within `domain_settings_initialize()` where it made no difference so well done for identifying that needed to go here. Now you've done that it makes sense.

I have suggested a change to match the corresponding part in https://github.com/backdrop/backdrop/pull/4567 and also raised a couple of other queries.  This will work even before the change is merged into core but will be ready from day 1 when people start testing the pre-release.

File: domain_settings/config/domain_settings.settings.json
Comment (rudy880719) [reply domain_settings/config/domain_settings.settings.json:1]: @yorkshire-pudding Yes, it is relevant, it has to do with the form configuration for each domain, specifically the domain settings module, take advantage and send those two changes

**Comment by rudy880719** [comment]
@yorkshire-pudding thanks for the help, I appreciate it, your changes were commited

**Review by yorkshire-pudding — APPROVED** [review]
Thanks @rudy880719 - looking good.

**Pull request merged**